### PR TITLE
Added gorilla mux router

### PR DESCRIPTION
**Pull request by bakurits**
#### Summary

- Added mux router 
- Deleted repetitive code and added middleware

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-github/issues/246

### CODE REVIEW

**Comment by bakurits** [comment]
@hanzei 
I cant found which HTTP methods are needed there

- /webhook
- /oauth/connect
- /oauth/complete
- /api/v1/todo

Also, I've added a few unit tests to test unauthorized behavior, I can commit those if needed

File: server/api.go [hunk 1]
```diff
@@ -61,216 +112,178 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
...
+				"github_username":     userInfo.GitHubUsername,
+				"github_client_id":    config.GitHubOAuthClientID,
+				"enterprise_base_url": config.EnterpriseBaseURL,
+				"organization":        config.GitHubOrg,
+			},
+			&model.WebsocketBroadcast{UserId: state.UserID},
+		)
+
+		html := `
+	<!DOCTYPE html>
```
Comment (larkox) [inline server/api.go:270]: nit: Here you have one indentation too many for this literal block.

File: server/api.go [hunk 2]
```diff
@@ -45,12 +50,58 @@ type PRDetails struct {
...
+	apiRouter.HandleFunc("/yourassignments", p.extractUserMiddleWare(p.getYourAssignments, false)).Methods("GET")
+	apiRouter.HandleFunc("/createissuecomment", p.extractUserMiddleWare(p.createIssueComment, false)).Methods("POST")
+	apiRouter.HandleFunc("/mentions", p.extractUserMiddleWare(p.getMentions, false)).Methods("GET")
+	apiRouter.HandleFunc("/unreads", p.extractUserMiddleWare(p.getUnreads, false)).Methods("GET")
+	apiRouter.HandleFunc("/settings", p.extractUserMiddleWare(p.updateSettings, false))
+	apiRouter.HandleFunc("/user", p.extractUserMiddleWare(p.getGitHubUser, true)).Methods("POST")
+}
+
+func (p *Plugin) extractUserMiddleWare(handler HTTPHandlerFuncWithUser, jsonResponse bool) http.HandlerFunc {
+
```
Comment (larkox) [inline server/api.go:88]: nit: I would remove this empty line and the one at the end of the block.

File: server/api.go (hunk 1 above)
Comment (bakurits) [reply server/api.go:270]: Do you mean this?
```go
html := `
	<!DOCTYPE html>
	<html>
		<head>
			<script>
				window.close();
			</script>
		</head>
		<body>
			<p>Completed connecting to GitHub. Please close this window.</p>
		</body>
	</html>
			`
```
or I should remove indications inside string literal

File: go.mod [hunk 3]
```diff
@@ -5,6 +5,7 @@ go 1.12
 require (
 	github.com/Masterminds/sprig/v3 v3.0.2
 	github.com/google/go-github/v25 v25.1.1
+	github.com/gorilla/mux v1.7.3
```
Comment (hanzei) [inline go.mod:8]: Nit pick: Why not use [`v1.7.4`](https://github.com/gorilla/mux/releases)?

File: server/api.go [hunk 4]
```diff
@@ -45,12 +50,58 @@ type PRDetails struct {
...
 	w.WriteHeader(err.StatusCode)
 	w.Write(b)
 }
 
+func (p *Plugin) initialiseAPI() {
+	p.router = mux.NewRouter()
+
+	webHookRouter := p.router.PathPrefix(API_WEBHOOK).Subrouter()
+	oauthRouter := p.router.PathPrefix(API_OAUTH).Subrouter()
+	apiRouter := p.router.PathPrefix(API_PREFIX).Subrouter()
```
Comment (hanzei) [inline server/api.go:66]: I'm unsure if it's really needed to have the routes be defined as global `const`. Is this needed for testing?

File: server/api.go [hunk 5]
```diff
@@ -45,12 +50,58 @@ type PRDetails struct {
...
+
+	webHookRouter.HandleFunc("/", p.handleWebhook)
+
+	oauthRouter.HandleFunc("/connect", p.extractUserMiddleWare(p.connectUserToGitHub, false))
+	oauthRouter.HandleFunc("/complete", p.extractUserMiddleWare(p.completeConnectUserToGitHub, false))
+
+	apiRouter.HandleFunc("/connected", p.extractUserMiddleWare(p.getConnected, true)).Methods("GET")
+	apiRouter.HandleFunc("/todo", p.extractUserMiddleWare(p.postToDo, true))
```
Comment (hanzei) [inline server/api.go:74]: Why does this method lack a `http method`?

File: server/api.go [hunk 6]
```diff
@@ -45,12 +50,58 @@ type PRDetails struct {
...
+	apiRouter.HandleFunc("/reviews", p.extractUserMiddleWare(p.getReviews, false)).Methods("GET")
+	apiRouter.HandleFunc("/yourprs", p.extractUserMiddleWare(p.getYourPrs, false)).Methods("GET")
+	apiRouter.HandleFunc("/prsdetails", p.extractUserMiddleWare(p.getPrsDetails, false)).Methods("POST")
+	apiRouter.HandleFunc("/searchissues", p.extractUserMiddleWare(p.searchIssues, false)).Methods("GET")
+	apiRouter.HandleFunc("/yourassignments", p.extractUserMiddleWare(p.getYourAssignments, false)).Methods("GET")
+	apiRouter.HandleFunc("/createissuecomment", p.extractUserMiddleWare(p.createIssueComment, false)).Methods("POST")
+	apiRouter.HandleFunc("/mentions", p.extractUserMiddleWare(p.getMentions, false)).Methods("GET")
+	apiRouter.HandleFunc("/unreads", p.extractUserMiddleWare(p.getUnreads, false)).Methods("GET")
+	apiRouter.HandleFunc("/settings", p.extractUserMiddleWare(p.updateSettings, false))
```
Comment (hanzei) [inline server/api.go:83]: Same question here

Comment (hanzei) [inline server/api.go:83]: Same question here

File: server/api.go [hunk 7]
```diff
@@ -45,12 +50,58 @@ type PRDetails struct {
 	Reviews            []*github.PullRequestReview `json:"reviews"`
 }
 
+type HTTPHandlerFuncWithUser func(w http.ResponseWriter, r *http.Request, userID string) http.HandlerFunc
```
Comment (hanzei) [inline server/api.go:53]: Why does this need to return a `http.HandlerFunc`? Is ` func(w http.ResponseWriter, r *http.Request, userID string)` as signature enough?

File: server/api.go [hunk 8]
```diff
@@ -45,12 +50,58 @@ type PRDetails struct {
...
+		if userID != "" {
+			handler(w, r, userID)(w, r)
+			return
+		}
+
+		if jsonResponse {
+			writeAPIError(w, &APIErrorResponse{ID: "", Message: "Not authorized.", StatusCode: http.StatusUnauthorized})
+		} else {
+			http.Error(w, "Not authorized", http.StatusUnauthorized)
+		}
```
Comment (hanzei) [inline server/api.go:100]: I'm 1/5 that the logic should be inverted given that https://github.com/mattermost/mattermost-plugin-github/pull/248/files#diff-0d5a44e9b96f048fa52b477ce6be6253R96-R100 is the error handling.

File: server/api.go (hunk 1 above)
Comment (larkox) [reply server/api.go:270]: The ones inside the string literal.

File: server/api.go (hunk 5 above)
Comment (hanzei) [reply server/api.go:74]: I missed your comment at https://github.com/mattermost/mattermost-plugin-github/pull/248#issuecomment-615868851.

- `/webhook` uses `POST`
- `/oauth/connect` uses `GET`
- `/oauth/complete` uses `GET`
- `/api/v1/todo` uses `POST`

**Comment by hanzei** [comment]
@bakurits Having more unit test would be great :+1:

File: server/api.go (hunk 5 above)
Comment (bakurits) [reply server/api.go:74]: @hanzei sorry I forgot `/api/v1/settings` is it `GET`?

**Comment by bakurits** [comment]
@hanzei I've added utility for test ServeHttp. For now, it only tests not authorized request types, because I'm not sure about endpoints behavior.

File: server/api.go [hunk 9]
```diff
@@ -45,12 +46,56 @@ type PRDetails struct {
 	Reviews            []*github.PullRequestReview `json:"reviews"`
 }
 
+type HTTPHandlerFuncWithUser func(w http.ResponseWriter, r *http.Request, userID string) func()
```
Comment (hanzei) [inline server/api.go:49]: Can this be simplified even more to  
```suggestion
type HTTPHandlerFuncWithUser func(w http.ResponseWriter, r *http.Request, userID string)
```
?

Then https://github.com/bakurits/mattermost-plugin-github/blob/b4f906611637384984d941bdfcf6ca81b0537755/server/api.go#L112-L114 could also be simplified to 
```go
func (p *Plugin) connectUserToGitHub(w http.ResponseWriter, r *http.Request, userID string) {
	privateAllowed := false
```

File: server/api_test.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestPlugin_ServeHTTP(t *testing.T) {
+
+	tassert := assert.New(t)
```
Comment (hanzei) [inline server/api_test.go:14]: Nit:
```suggestion
func TestPlugin_ServeHTTP(t *testing.T) {
	tassert := assert.New(t)
```

File: server/util/util.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+	// ContentTypeHTML string associated with content-type html
+	ContentTypeHTML ContentType = "text/html"
+
+	// ContentTypePlain
+	ContentTypePlain ContentType = "text/plain; charset=utf-8"
+)
+
+// ContentType type for http Content-Type
+type ContentType string
```
Comment (hanzei) [inline server/util/util.go:25]: Nit: Does the `type` need go be exported?

File: server/util/util.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+	req, err := http.NewRequest(request.Method, request.URL, body)
+	test.NoError(err, "Error while creating Request")
+	return req
+}
+
+// CompareHTTPResponse compares expected response with real one
+func (test *HTTPTest) CompareHTTPResponse(rr *httptest.ResponseRecorder, expected ExpectedResponse) {
+
+	test.Equal(expected.StatusCode, rr.Code, "Http status codes are different")
```
Comment (hanzei) [inline server/util/util.go:89]:
```suggestion
func (test *HTTPTest) CompareHTTPResponse(rr *httptest.ResponseRecorder, expected ExpectedResponse) {
	test.Equal(expected.StatusCode, rr.Code, "Http status codes are different")
```

File: server/util/util.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
+package util
```
Comment (hanzei) [inline server/util/util.go:1]: :+1: for having this in a separate package. Maybe `testutils` is a better package name? Then this file could be `http.go`.

File: server/util/util.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+// ExpectedResponse stores expected responce basic data
+type ExpectedResponse struct {
+	StatusCode   int
+	ResponseType ContentType
+	Body         interface{}
+}
+
+// HTTPTest encapsulates data for testing needs
+type HTTPTest struct {
+	*assert.Assertions
```
Comment (hanzei) [inline server/util/util.go:43]: I leaning towards just passing `*testing.T` here, which would be more generic. Thoughts?

File: server/util/util.go (hunk 11 above)
Comment (bakurits) [reply server/util/util.go:25]: I'm using it [here](https://github.com/bakurits/mattermost-plugin-github/blob/b4f906611637384984d941bdfcf6ca81b0537755/server/api_test.go#L43)

File: server/api.go (hunk 5 above)
Comment (hanzei) [reply server/api.go]: Yes

File: server/util/util.go (hunk 11 above)
Comment (hanzei) [reply server/util/util.go:25]: Surely `ContentTypeJSON` need to be exported. But does `ContentType`?

Comment (bakurits) [reply server/util/util.go:25]: Ah sorry for the misunderstanding. I'll change that

**Comment by hanzei** [comment]
@DHaussermann There is no specific area to test here, but every part of the plugin should be smoke test specially (dis)connecting a user.

**Comment by DHaussermann** [comment]
@hanzei If you have a bit of time, would you like a try out the peer testing process documented here https://mattermost.atlassian.net/wiki/spaces/INT/pages/619937892/Peer+Testing+Process on this PR?
If you have a chance, I would hugely appreciate your feedback on what could be clarified further.

**Comment by hanzei** [comment]
@bakurits Heads up that you have to merge `master` into your branch because of https://github.com/mattermost/mattermost-plugin-github/pull/250

**Comment by bakurits** [comment]
@hanzei done  👌

**Comment by hanzei** [comment]
@bakurits :+1: Could you please look into the issue with the webhook that I found?

**Comment by bakurits** [comment]
```go
switch path := r.URL.Path; path {		
	case "/webhook":
		p.handleWebhook(w, r)
```
there wasn't `/` in the end previously

**Comment by bakurits** [comment]
@hanzei you were right `handleWebhook` wasn't being called from unit tests either. Not it is, can you recheck from Github?

File: server/api.go [hunk 15]
```diff
@@ -45,12 +46,56 @@ type PRDetails struct {
...
 }
 
+func (p *Plugin) initialiseAPI() {
+	p.router = mux.NewRouter()
+
+	webHookRouter := p.router.PathPrefix("/webhook").Subrouter()
+	oauthRouter := p.router.PathPrefix("/oauth").Subrouter()
+	apiRouter := p.router.PathPrefix("/api/v1").Subrouter()
+
+	webHookRouter.HandleFunc("", p.handleWebhook).Methods("POST")
```
Comment (hanzei) [inline server/api.go:64]: With this change, is a subrouter really needed?

Comment (bakurits) [reply server/api.go:64]: No, I think it's redundant now. I'll change that

**Comment by hanzei** [comment]
@DHaussermann :point_up:

**Comment by hanzei** [comment]
Thanks again for the great code improvement @bakurits :tada:

**Pull request merged**